### PR TITLE
Enable SD XL ONNX export and ONNX Runtime inference

### PR DESCRIPTION
**Pull request by echarlaix**
Add Stable Diffusion XL ONNX export and pipelines to enable ONNX Runtime inference for text-to-image and image-to-image 

Export using the CLI :
```
optimum-cli export onnx --model stabilityai/stable-diffusion-xl-base-0.9 --task stable-diffusion-xl ./sd_xl_onnx
```
 
For inference :
```python
from optimum.onnxruntime import ORTStableDiffusionXLPipeline

model_id = "stabilityai/stable-diffusion-xl-base-0.9"
pipeline = ORTStableDiffusionXLPipeline.from_pretrained(model_id, export=True)
```

### CODE REVIEW

File: optimum/onnxruntime/modeling_diffusion.py [hunk 1]
```diff
@@ -43,9 +43,8 @@
 from ..pipelines.diffusers.pipeline_stable_diffusion_img2img import StableDiffusionImg2ImgPipelineMixin
 from ..pipelines.diffusers.pipeline_stable_diffusion_inpaint import StableDiffusionInpaintPipelineMixin
 from ..pipelines.diffusers.pipeline_stable_diffusion_xl import StableDiffusionXLPipelineMixin
-from ..pipelines.diffusers.pipeline_stable_diffusion_xl_img2img import StableDiffusionXLImg2ImgPipelineMixin
-
 
+# from ..pipelines.diffusers.pipeline_stable_diffusion_xl_img2img import StableDiffusionXLImg2ImgPipelineMixin
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_diffusion.py:47]: Should this be removed?

File: optimum/onnxruntime/modeling_diffusion.py [hunk 2]
```diff
@@ -590,6 +589,8 @@ def __call__(self, *args, **kwargs):
         return StableDiffusionXLPipelineMixin.__call__(self, *args, **kwargs)
 
 
+"""
 class ORTStableDiffusionXLImg2ImgPipeline(ORTStableDiffusionXLPipelineBase, StableDiffusionXLImg2ImgPipelineMixin):
     def __call__(self, *args, **kwargs):
         return StableDiffusionXLImg2ImgPipelineMixin.__call__(self, *args, **kwargs)
+"""
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_diffusion.py]: Same

**Review by fxmarty — COMMENTED** [review]
Looks good, thanks for working on it! Probably we could have export / run tests & doc as well in the PR?

File: optimum/exporters/tasks.py [hunk 3]
```diff
@@ -931,7 +937,14 @@ class TasksManager:
             onnx="YolosOnnxConfig",
         ),
     }
-    _UNSUPPORTED_CLI_MODEL_TYPE = {"unet", "vae-encoder", "vae-decoder", "clip-text-model", "trocr"}
+    _UNSUPPORTED_CLI_MODEL_TYPE = {
```
Comment (fxmarty) [inline optimum/exporters/tasks.py:940]: Could you add an explicit check in `main_export` with meaningful error?

File: optimum/onnxruntime/modeling_diffusion.py (hunk 1 above)
Comment (echarlaix) [reply optimum/onnxruntime/modeling_diffusion.py:47]: I was planning to add it in this PR (will uncomment when ready)

File: optimum/onnxruntime/modeling_diffusion.py [hunk 4]
```diff
@@ -494,3 +543,54 @@ def __call__(self, *args, **kwargs):
...
+class ORTStableDiffusionXLPipeline(ORTStableDiffusionXLPipelineBase, StableDiffusionXLPipelineMixin):
+    def __call__(self, *args, **kwargs):
+        return StableDiffusionXLPipelineMixin.__call__(self, *args, **kwargs)
+
+
+"""
+class ORTStableDiffusionXLImg2ImgPipeline(ORTStableDiffusionXLPipelineBase, StableDiffusionXLImg2ImgPipelineMixin):
+    def __call__(self, *args, **kwargs):
+        return StableDiffusionXLImg2ImgPipelineMixin.__call__(self, *args, **kwargs)
+"""
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_diffusion.py:596]: To be removed?

File: optimum/pipelines/diffusers/watermark.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import numpy as np
+from imwatermark import WatermarkEncoder
+
+
+WATERMARK_MESSAGE = 0b101100111110110010010000011110111011000110011110
+WATERMARK_BITS = [int(bit) for bit in bin(WATERMARK_MESSAGE)[2:]]
+
+
+# Copied from https://github.com/huggingface/diffusers/blob/v0.18.1/src/diffusers/pipelines/stable_diffusion_xl/watermark.py#L12
```
Comment (fxmarty) [inline optimum/pipelines/diffusers/watermark.py:9]: Why not just import those from diffusers instead of c/c?

File: optimum/utils/normalized_config.py [hunk 6]
```diff
@@ -30,7 +30,7 @@ class NormalizedConfig:
     """
 
     def __init__(self, config: Union[PretrainedConfig, Dict], allow_new: bool = False, **kwargs):
-        self.config = config if isinstance(config, PretrainedConfig) else PretrainedConfig.from_dict(config)
+        self.config = config
```
Comment (fxmarty) [inline optimum/utils/normalized_config.py:33]: Why?

File: optimum/pipelines/diffusers/pipeline_stable_diffusion_xl_img2img.py [hunk 7]
```diff
@@ -0,0 +1,503 @@
...
+            latents = generator.randn(*shape).astype(dtype)
+        elif latents.shape != shape:
+            raise ValueError(f"Unexpected latents shape, got {latents.shape}, expected {shape}")
+
+        # scale the initial noise by the standard deviation required by the scheduler
+        latents = latents * np.float64(self.scheduler.init_noise_sigma)
+
+        return latents
+
+    # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline.prepare_extra_step_kwargs
```
Comment (fxmarty) [inline optimum/pipelines/diffusers/pipeline_stable_diffusion_xl_img2img.py:251]: Couldn't it make sense to inherit from diffusers pipelines to avoid many of the copied from? I feel like we already discussed this but forgot the motivation not to?

File: optimum/exporters/tasks.py [hunk 8]
```diff
@@ -171,6 +171,7 @@ class TasksManager:
             "audio-xvector": "AutoModelForAudioXVector",
             "image-to-text": "AutoModelForVision2Seq",
             "stable-diffusion": "StableDiffusionPipeline",
+            "stable-diffusion-xl": "StableDiffusionXLPipeline",
```
[the remark this answers is not in the document]
Comment (echarlaix) [reply optimum/exporters/tasks.py:174]: Agreed!

File: optimum/onnxruntime/modeling_diffusion.py (hunk 4 above)
Comment (echarlaix) [reply optimum/onnxruntime/modeling_diffusion.py]: Done in [`e5d0bb`](https://github.com/huggingface/optimum/pull/1168/commits/e5d0bb4b2e6bc6da035e1c126a6c57b1c572dc30)

File: optimum/pipelines/diffusers/pipeline_stable_diffusion_xl_img2img.py (hunk 7 above)
Comment (echarlaix) [reply optimum/pipelines/diffusers/pipeline_stable_diffusion_xl_img2img.py]: This actually concerns only the `check_inputs` method (see updates) so I don't think it would make sense to inherit from diffusers pipelines for this specifically.

File: optimum/utils/normalized_config.py (hunk 6 above)
Comment (echarlaix) [reply optimum/utils/normalized_config.py:33]: We can't create an instance of `PretrainedConfig` for certain `diffusers` config since the addition of https://github.com/huggingface/diffusers/blob/8d8b4311b9e19c3621dbd69e6690ee7be5702f65/src/diffusers/configuration_utils.py#L553 [`_use_default_values`](https://github.com/huggingface/diffusers/blob/8d8b4311b9e19c3621dbd69e6690ee7be5702f65/src/diffusers/configuration_utils.py#L610) which is a set (so not JSON serializable). We could easily make it compatible, but we shouldn't even create an instance of `PretrainedConfig` in the first place IMO, it's not clear to me why `transformers` should be a hard dependency when a user wants to export a `diffusers` model

File: optimum/pipelines/diffusers/pipeline_stable_diffusion_xl_img2img.py (hunk 7 above)
Comment (fxmarty) [reply optimum/pipelines/diffusers/pipeline_stable_diffusion_xl_img2img.py]: I see, I was mislead by the Copied from

File: optimum/pipelines/diffusers/pipeline_stable_diffusion_xl.py [hunk 9]
```diff
@@ -0,0 +1,499 @@
...
+        self,
+        prompt: Union[str, List[str]],
+        height: Optional[int],
+        width: Optional[int],
+        callback_steps: int,
+        negative_prompt: Optional[str] = None,
+        prompt_embeds: Optional[np.ndarray] = None,
+        negative_prompt_embeds: Optional[np.ndarray] = None,
+        pooled_prompt_embeds=None,
+        negative_pooled_prompt_embeds=None,
```
Comment (michaelbenayoun) [inline optimum/pipelines/diffusers/pipeline_stable_diffusion_xl.py:179]:
```suggestion
        pooled_prompt_embeds: Optional[] =None,
        negative_pooled_prompt_embeds: Optional[]=None,
```

File: tests/exporters/onnx/test_exporters_onnx_cli.py [hunk 10]
```diff
@@ -256,6 +256,14 @@ def test_stable_diffusion(self):
                 check=True,
             )
 
+    def test_stable_diffusion_xl(self):
+        with TemporaryDirectory() as tmpdirname:
+            subprocess.run(
+                f"python3 -m optimum.exporters.onnx --model echarlaix/tiny-random-stable-diffusion-xl --task stable-diffusion-xl {tmpdirname}",
+                shell=True,
+                check=True,
+            )
```
[the remark this answers is not in the document]
Comment (echarlaix) [reply tests/exporters/onnx/test_exporters_onnx_cli.py]: Done !

File: optimum/exporters/tasks.py (hunk 3 above)
Comment (echarlaix) [reply optimum/exporters/tasks.py:940]: done in [`6c122`](https://github.com/huggingface/optimum/pull/1168/commits/6c12200d33e7c9451fc1afa59f6251aeed6c3205)

File: optimum/exporters/onnx/__main__.py [hunk 11]
```diff
@@ -415,28 +410,24 @@ def main_export(
...
 
-        onnx_files_subpaths = [os.path.join(path, ONNX_WEIGHTS_NAME) for path in onnx_files_subpaths]
+        onnx_files_subpaths = [os.path.join(name_dir, ONNX_WEIGHTS_NAME) for name_dir in models_and_onnx_configs]
 
         # Saving the additional components needed to perform inference.
-        model.tokenizer.save_pretrained(output.joinpath("tokenizer"))
         model.scheduler.save_pretrained(output.joinpath("scheduler"))
-        if model.feature_extractor is not None:
+        if getattr(model, "feature_extractor", None) is not None:
             model.feature_extractor.save_pretrained(output.joinpath("feature_extractor"))
```
[the remark this answers is not in the document]
Comment (echarlaix) [reply optimum/exporters/onnx/__main__.py:430]: Here we also need to verify that its value is not `None` so second option wouldn't work as is, will update with first

File: optimum/exporters/onnx/model_configs.py [hunk 12]
```diff
@@ -677,13 +678,33 @@ def inputs(self) -> Dict[str, Dict[int, str]]:
...
-        return {
+        outputs = {
+            "text_embeds": {0: "batch_size", 1: "sequence_length"},
+            "last_hidden_state": {0: "batch_size", 1: "sequence_length"},
+        }
+        if self._normalized_config.output_hidden_states:
+            for i in range(self._normalized_config.num_layers + 1):
+                outputs[f"hidden_states.{i}"] = {0: "batch_size", 1: "sequence_length"}
+
+        return outputs
```
[the remark this answers is not in the document]
Comment (echarlaix) [reply optimum/exporters/onnx/model_configs.py:689]: Yes the name of the property is "outputs" but it won't have any impact, will update the variable name for clarity !

File: optimum/exporters/onnx/model_configs.py [hunk 13]
```diff
@@ -734,8 +762,18 @@ def torch_to_onnx_output_map(self) -> Dict[str, str]:
...
+        if getattr(self._normalized_config, "addition_embed_type", None) == "text_time":
+            dummy_inputs["added_cond_kwargs"] = {
+                "text_embeds": dummy_inputs.pop("text_embeds"),
+                "time_ids": dummy_inputs.pop("time_ids"),
+            }
+
         return dummy_inputs
 
+    def ordered_inputs(self, model) -> Dict[str, Dict[int, str]]:
+        return self.inputs
```
[the remark this answers is not in the document]
Comment (echarlaix) [reply optimum/exporters/onnx/model_configs.py]: Here there is a mismatch between the pytorch model forward signature and the ONNX model expected inputs :  `added_cond_kwargs` is a dictionnary that depends on `config.addition_embed_type` and that needs to be unpacked. Just added an update in [`c52284`](https://github.com/huggingface/optimum/pull/1168/commits/c5228405fa8bad76117f7c7f307152340cd2a180) with comment to explain

File: optimum/utils/normalized_config.py (hunk 6 above)
Comment (echarlaix) [reply optimum/utils/normalized_config.py]: Fixed in https://github.com/huggingface/diffusers/pull/4040, still think we don't need to create a `PretrainedConfig` instance

File: tests/exporters/onnx/test_onnx_export.py [hunk 14]
```diff
@@ -314,6 +311,30 @@ def _onnx_export(
 
                 gc.collect()
 
+    def _onnx_export_sd(self, model_type: str, model_name: str, device="cpu"):
```
Comment (fxmarty) [inline tests/exporters/onnx/test_onnx_export.py:314]: I think the only benefit of this `test_onnx_export.py` is to test on various shapes the exported model.

Couldn't we just remove those tests for SD? Does it bring anything more than the tests you put in `test_exporters_onnx_cli`?

**Review by fxmarty — APPROVED** [review]
LGTM, awesome!

File: tests/exporters/onnx/test_onnx_export.py (hunk 14 above)
Comment (echarlaix) [reply tests/exporters/onnx/test_onnx_export.py:314]: In `test_onnx_export.py` we verify whether each model's subcomponent is correctly exported by comparing the output of the original subcomponent with the exported one : https://github.com/huggingface/optimum/blob/v1.9.1/tests/exporters/onnx/test_onnx_export.py#L411

File: optimum/exporters/onnx/__main__.py [hunk 15]
```diff
@@ -415,28 +416,31 @@ def main_export(
...
+            feature_extractor.save_pretrained(output.joinpath("feature_extractor"))
+
+        tokenizer = getattr(model, "tokenizer", None)
+        if tokenizer is not None:
+            tokenizer.save_pretrained(output.joinpath("tokenizer"))
+
+        tokenizer_2 = getattr(model, "tokenizer_2", None)
+        if tokenizer_2 is not None:
+            tokenizer_2.save_pretrained(output.joinpath("tokenizer_2"))
+
```
[the remark this answers is not in the document]
Comment (echarlaix) [reply optimum/exporters/onnx/__main__.py:443]: I'm not sure `maybe_save_preprocessors` is adapted here as we would need to load the different preprocessors a second time (already done when instantiating a pipeline), happy to add this integration in a second PR if  you have a specific idea (maybe I'm missing something)

File: tests/onnxruntime/test_stable_diffusion_pipeline.py [hunk 16]
```diff
@@ -222,6 +229,71 @@ def test_image_reproducibility(self, model_arch: str):
...
+            "prompt": "sailing ship in storm by Leonardo da Vinci",
+            "num_inference_steps": 1,
+            "num_images_per_prompt": num_images_per_prompt,
+            "height": height,
+            "width": width,
+            "guidance_rescale": 0.1,
+        }
+
+        for output_type in ["latent", "np"]:
+            ort_outputs = ort_pipeline(latents=latents, output_type=output_type, **kwargs).images
```
[the remark this answers is not in the document]
Comment (echarlaix) [reply tests/onnxruntime/test_stable_diffusion_pipeline.py:271]: Good point, adapted the test to check for batch size > 1 in [`0426f`](https://github.com/huggingface/optimum/pull/1168/commits/0426fa446f065706352ee5d9b1dca66aa5414104)

File: optimum/exporters/onnx/__main__.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (fxmarty) [reply optimum/exporters/onnx/__main__.py:443]: Not necessarily, it could be possible to pass the model/pipeline. It's just to keep a similar style as for transformers, not important.

File: tests/exporters/onnx/test_exporters_onnx_cli.py [hunk 17]
```diff
@@ -134,6 +135,31 @@ def test_all_models_tested(self):
...
+    @require_diffusers
+    def test_exporters_cli_pytorch_cpu_stable_diffusion(self, model_type: str, model_name: str):
+        self._onnx_export(model_name, model_type)
+
+    @parameterized.expand(PYTORCH_STABLE_DIFFUSION_MODEL.items())
+    @require_torch_gpu
+    @require_vision
+    @require_diffusers
+    @slow
+    @pytest.mark.run_slow
```
Comment (michaelbenayoun) [inline tests/exporters/onnx/test_exporters_onnx_cli.py:150]: Why?
Maybe I forgot the reason for this.

**Review by michaelbenayoun — APPROVED** [review]
LGTM, thanks @echarlaix

File: tests/exporters/onnx/test_exporters_onnx_cli.py (hunk 17 above)
Comment (echarlaix) [reply tests/exporters/onnx/test_exporters_onnx_cli.py:150]: triggered everyday by [`test_exporters_slow.yml`](https://github.com/huggingface/optimum/blob/c4750f6014d8ac39bca780e1e4381b15eaab9197/.github/workflows/test_exporters_slow.yml) (not all tests are triggered for every PR)

**Comment by NeusZimmer** [comment]
Any option to run fp16 without cuda for O4 optimization on AMD?
